### DISCUSSION TITLE
Lottie turns out to be 300MB+ in size when fetched through SPM

### DISCUSSION THREAD

**Discussion by AvdLee** [airbnb/lottie-ios#1807] [General]
## Which Version of Lottie are you using?

Lottie 3.5.0

## Expected Behavior 
Lottie should not be so large.

## Actual Behavior
Lottie is fetched as a 357MB size repository somehow:

```bash
avanderlee@Antoines-MacBook-Pro-2 repositories % du -sh */ | sort -hr
357M	lottie-ios-0271592b/
173M	firebase-ios-sdk-a1569f98/
117M	grpc-ios-9a5f03c1/
.....
```

It's mostly due to this file:
![CleanShot 2022-11-10 at 15 19 49@2x](https://user-images.githubusercontent.com/4329185/201116126-91134a6d-63a0-4d77-86d4-c95e08019b55.jpg)

I'm not sure yet how this is possible since the repository currently downloaded is only ~150MB.

Any input is welcome!

**Comment by AvdLee** [comment]
It might be related to GIT history: https://stackoverflow.com/questions/11050265/remove-large-pack-file-created-by-git

**Comment by calda** [comment]
A clean clone of the repo is currently 315MB. This is mostly because we have ~100MB of snapshot images from our regression tests checked into the repo (`Tests/__Snapshots__/SnapshotTests`), plus their git history.

**Reply by calda** [reply]
We may be able to move the snapshots to [LFS](https://docs.github.com/en/repositories/working-with-files/managing-large-files/about-git-large-file-storage), I'll see if this is feasible

**Reply by calda** [reply]
LFS is super easy to integrate into a repo, but the main issue is that we'd lose the ability to view snapshot diffs in pull requests: https://docs.github.com/en/repositories/working-with-files/managing-large-files/collaboration-with-git-large-file-storage

We'd need some sort of replacement flow for viewing the snapshot diffs caused by a pull request, which is important for reviewing and understanding the changing in a PR.

**Reply by AvdLee** [reply]
Gotya, that clarifies! So one thing you could do is adding support for binary targets. This is a repo that did it for Firebase: https://github.com/akaffenberger/firebase-ios-sdk-xcframeworks which can be inspiring, but it mostly comes down to providing a binary release.

For example, this is how we reference SwiftLint through a binary target:

```swift
.binaryTarget(
    name: "SwiftLintBinary",
    url: "https://github.com/realm/SwiftLint/releases/download/0.49.1/SwiftLintBinary-macos.artifactbundle.zip",
    checksum: "227258fdb2f920f8ce90d4f08d019e1b0db5a4ad2090afa012fd7c2c91716df3"
)
```

The benefit is that we only have to download that artifact, only the files we really need. It might also have an impact on build times, so it could definitely be something to explore.

**Reply by calda** [reply]
We now publish `xcframework` archives as a part of our release process: https://github.com/airbnb/lottie-ios/discussions/1807#discussioncomment-4346438

**Comment by calda** [comment]
We've adopted Git LFS for our snapshot images. By default a checkout of master is now 213MB, down from 375MB.

**Reply by AvdLee** [reply]
That's still a lot compared to the binary framework, which is only 9MB. We've adopted it ourselves and decided to create them manually, but it would be great if that could be supported by Lottie natively until we have a package registry available.

**Reply by SajjadKharrazi** [reply]
i think lottie can have an approach like Branch.io

https://github.com/BranchMetrics/ios-branch-deep-linking-attribution/issues/1053

**Reply by AvdLee** [reply]
A separate repo sounds quite extreme to me! Just providing artifact bundles to the release should be good enough and a significant step forward already. [SwiftLint recently did improve this support as well.](https://github.com/realm/SwiftLint/pull/4603).

**Reply by calda** [reply]
@AvdLee, how involved was the process of manually creating and adopting a binary framework? Depending on the complexity, it may make sense for us to document how to do this for folks who are interested (and concerned about the size of the repo as-is).

**Reply by AvdLee** [reply]
It was doable for me since I'm knowledgeable in the area. However, I believe it should be something Lottie should offer in return for the large repository size. It shouldn't be too complex to write an automated script for it that you can add to the release process. It makes your repository look better and helps the many implementors of this framework.

Ensure to learn from other frameworks providing similar `xcframework` alternatives.

**Reply by xmollv** [reply]
> That's still a lot compared to the binary framework, which is only 9MB. We've adopted it ourselves and decided to create them manually, but it would be great if that could be supported by Lottie natively until we have a package registry available.

This would be awesome. Lottie is by far the slowest dependency to pull in our case. I've updated to 4.0.0 this morning and SPM reports that it takes a bit more than 2 minutes for Lottie.

<img width="445" alt="Screenshot 2022-12-07 at 10 28 58" src="https://user-images.githubusercontent.com/8229382/206141107-c3c9d774-92dc-47b7-9c14-f712914782f9.png">

**Comment by calda** [comment]
We now publish `xcframework` archives as a part of our release process: https://github.com/airbnb/lottie-ios/pull/1843#issue-1485294198

For example, the [4.0.1 release](https://github.com/airbnb/lottie-ios/releases/tag/4.0.1) includes a `Lottie.xcarchive.zip` attachhment. You can install this into your Xcode project manually, or integrate it using Swift Package Manager:

```swift
// You can either use a remote binary dependency where SPM downloads the xcframework from GitHub
.binaryTarget(
  name: "Lottie",
  url: "https://github.com/airbnb/lottie-ios/releases/download/4.0.1/Lottie.xcframework.zip",
  checksum: "b6d8b0b81975d91965b8bb00cffb0eae4b3d94538b6950a90bc1366afd5d4239")

// Or you can download the xcframework locally and include it in your project's git repo:
.binaryTarget(
  name: "Lottie",
  path: "path/to/Lottie.xcframework")
```

The `xframework` is only 8.6MB.

**Reply by xmollv** [reply]
@calda That's awesome! Is there any way to make SPM -at the project level- default to use the `xcframework` instead of pulling the git repo? For example, Instabug [distributes](https://github.com/Instabug/Instabug-SP/blob/master/Package.swift) the SDK directly as an `xcframework` when pulled via SPM. Is this something that Lottie will support?

Your example seems to be towards embedding Lottie as a dependency to another SPM package, but I can't seem to find a way to make SPM pull just the `xcframework` when using it from the project itself. I've started using SPM recently, so I might be missing something obvious, sorry in advance if that's the case! 😬

**Reply by calda** [reply]
What do you mean by "at the project level"?

**Reply by xmollv** [reply]
@calda Sorry for the lack of better words, but I mean adding the dependency to the project, not to another package. In your example, you're defining the Lottie dependency in a `Package.swift`, and as far as I know, the `Package.swift` is only used by packages to define their dependencies, not by the project itself (by project I mean Xcode, screenshot below). For example, I've created a new project and included Lottie as a dependency via the GUI for SPM:

<img width="1434" alt="Screenshot 2022-12-09 at 15 49 18" src="https://user-images.githubusercontent.com/8229382/206728391-18411b92-4508-4586-9bce-0d380e57bcd4.png">

It pulls the entire repo as usual. My question is if there's any way to make it only pull the `xcframework` to reduce the time that it takes for the project to be fetched remotely (what I'm after is trying to speed up our CI build times basically). Here's how much it took to add it:
<img width="422" alt="Screenshot 2022-12-09 at 15 49 30" src="https://user-images.githubusercontent.com/8229382/206728620-07bdcab6-7234-4b01-85ab-a25e3d7e13e7.png">

**Reply by calda** [reply]
This doesn't appear to be supported by Xcode.

I'm exploring ways to simplify this, including publishing a separate "lottie-spm" repo that wraps the binary target in a way that's compatible with Xcode.

For example, [this proof of concept repo](https://github.com/calda/lottie-spm) can be added via Xcode like any other package, but actually just wraps Lottie's binary targets:

<img width="775" alt="image" src="https://user-images.githubusercontent.com/1811727/206771628-b79ffb11-774b-453c-af33-da2e351204fb.png">

```swift
let package = Package(
  name: "Lottie",
  products: [.library(name: "Lottie", targets: ["Lottie"])],
  targets: [
    .binaryTarget(
      name: "Lottie",
      url: "https://github.com/airbnb/lottie-ios/releases/download/4.0.1/Lottie.xcframework.zip",
      checksum: "b6d8b0b81975d91965b8bb00cffb0eae4b3d94538b6950a90bc1366afd5d4239"),
  ])
```

Obligatory disclaimer: the repo I linked is an unofficial proof of concept and not something to use in your project quite yet. I'm exploring how best to handle this, and will share an update once we know more.

**Reply by xmollv** [reply]
@calda I see. Maybe this is something that should be handled automatically by SPM when you're integrating it via Xcode, but that's something out of our hands I guess. Thanks for taking a look at this, even if it would be quite handy, I don't think that every SPM maintainer should have a separate repo just so that SPM can point to the binary distribution.

**Reply by AvdLee** [reply]
@calda this is amazing! We'll be migrating back to this soon. I'll let you know if we run into any issues!

**Reply by SajjadKharrazi** [reply]
@calda when i added .xcframework file manually and upload my app to appStore, i got this error:

```
ITMS-90863: Apple silicon Macs support issue - The app links with libraries that are not present on Mac:

@rpath/Lottie.framework/Lottie
After you’ve corrected the issues, you can upload a new binary to App Store Connect.
```

but without .xc everything is ok

**Reply by SajjadKharrazi** [reply]
also it causes app to crash when build on physical device, but on simulator it does not

**Reply by alexfringes** [reply]
Unsure if this is helpful, but we might be able to learn something from how @akaffenberger has set up this binary SPM mirror for the Firebase iOS sdk: https://github.com/akaffenberger/firebase-ios-sdk-xcframeworks
Specifically, this includes some project-level changes that seemingly are needed, as well as the script that creates the binary when the original repo releases a new version.

**Reply by calda** [reply]
That's what we're hoping to set up for Lottie -- I have an early proof-of-concept for this [here](https://github.com/calda/lottie-spm) and should be able to finish setting this up in an official repo early in January.

**Reply by alexfringes** [reply]
So glad to hear you’re considering rolling an official version. My reply was honestly more meant as a suggestion for the problems @SajjadKharrazi was encountering, which I do not run into with the Firebase binary mirror. E.g., I wonder if the -Objc flag is required atm?

**Comment by calda** [comment]
Thanks all for the discussion here. We've created a new [lottie-spm](https://github.com/airbnb/lottie-spm) repo that folks can use with Swift Package Manager. This repo is much smaller than the main repo (~500kb instead of 300mb) so should resolve these repo size issues. It's easier to integrate than XCFrameworks, since it's just a normal Swift package that you can add via Xcode or your `Package.swift`.

**Reply by xmollv** [reply]
@calda Thanks for doing this! We've moved to the SPM specific package and so far so good.

**Reply by Ernest0-Production** [reply]
Hi! The new repository began to grow inadequately in size :/ 

It has about 20 commits, but it weighs much more than other older  repositories like Alamofire, kingfisher

![image](https://github.com/airbnb/lottie-ios/assets/19843524/83df6fd4-733d-46c3-9754-2854bd17d88e)

**Reply by calda** [reply]
That's the .build directory, not just the source checkout. A fresh clone of the lottie-spm repo is only 200kb.

The .build directory probably also includes the XCFramework archives which are downloaded at build time, but not part of the git repository.

**Reply by Ernest0-Production** [reply]
No) That's repository folder 

<img width="793" alt="image" src="https://github.com/airbnb/lottie-ios/assets/19843524/371bfab7-ae18-43ba-b938-cc8d6de9af44">
